### DISCUSSION TITLE
etcd plugin behaviour and CNAME(s)

### DISCUSSION THREAD

**Discussion by prologic** [coredns/coredns#4785] [Q&A] [answered]
Hi All 👋

I'm James Mills (_obviously_) and I work for a company called AUCloud where I'm currently attempting to work on some internal DNS solutions. Being an IaaS provider I felt that having Terraform managed DNS internally was an important step in the right direction.

To that end I've been working on writing a [libdns](https://github.com/libdns) provider for CoreDNS by taking advantage of the wonderful [etcd plugin](https://coredns.io/plugins/etcd/).

I've gotten as far as supporting single A records and multiple A records as well as even supporting (_I think_) CNAME(s).

But I've run into some weird behaviours I'd like clarified (_maybe I'm doing something wrong_):

I'm seeing CoreDNS return the atual CNAME record itself as part of the ANSWER section.

See:

```#!sh
$ dig +short @127.0.0.1 -p 5300 -t A bar.test
foo.test.
1.2.3.4
5.6.7.8
```

And:

```#!sh
$ dig @127.0.0.1 -p 5300 -t A bar.test

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 5300 -t A bar.test
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 17320
;; flags: qr aa rd; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;bar.test.			IN	A

;; ANSWER SECTION:
bar.test.		300	IN	CNAME	foo.test.
foo.test.		300	IN	A	1.2.3.4
foo.test.		300	IN	A	5.6.7.8

;; Query time: 6 msec
;; SERVER: 127.0.0.1#5300(127.0.0.1)
;; WHEN: Mon Aug 09 15:39:40 AEST 2021
;; MSG SIZE  rcvd: 115
```

my [libdns-coredns](https://git.mills.io/prologic/libdns-coredns) implementation basically adds records to etcd (_currentlt_) like this:

```#!sh
$ etcdctl --endpoints=http://127.0.0.1:2379 get --prefix ""
/coredns/test/bar
{"host":"foo.test."}
/coredns/test/foo/A1
{"host":"1.2.3.4"}
/coredns/test/foo/A2
{"host":"5.6.7.8"}
```

Source code can be found at https://git.mills.io/prologic/libdns-coredns as well as the `cname` branch I'm working on. This is still very much work-in-rogress.

Appreciatee any helpful hints!

**Comment by chrisohaver** [comment, marked as answer]
> I'm seeing CoreDNS return the atual CNAME record itself as part of the ANSWER section.

This behavior for CNAMEs follows DNS RFC 1034.

**Reply by chrisohaver** [reply]
The behavior described (in step 3.a): https://datatracker.ietf.org/doc/html/rfc1034#section-4.3.2

And an example query/response:  https://datatracker.ietf.org/doc/html/rfc1034#section-6.2.7

**Reply by prologic** [reply]
Oh wow 😳 I guess I'm out-of-date with the DNS spec or has that always been the case?

**Reply by prologic** [reply]
Yeah looks like you're right, geez 😳 I'd better go re-read the DNS specs 😂 Sorry for the fud! I'll keep doing then since what I've built so far is _actually_ working and _behaving_ correctly.
